### PR TITLE
Add relic activation ability

### PR DESCRIPTION
**Pull request by anothersimulacrum**
#### Summary

SUMMARY: Features "Add activation abilities and charges to relics"

#### Purpose of change
Requested by @KorGgenT to further obsolete the artifact code.

#### Describe the solution
Encapsulate all the charge info about relics in a relic_charge_info struct - then allow specifying variations of this in the relic_procgen_data.

Add simple placeholder versions to this data, along with active effects for it to provide.

Attempt to recharge relics when processing them.

#### Testing
Spawn in a relic, activate it. See that it gives you pain. Wait for more than 6 hours, see that the relic has recharged.

#### Additional context
relic1
relic2

### CODE REVIEW

File: src/item.cpp [hunk 1]
```diff
@@ -4676,6 +4676,9 @@ std::string item::tname( unsigned int quantity, bool with_prefix, unsigned int t
     if( gunmod_find( itype_barrel_small ) ) {
         modtext += _( "sawn-off " );
     }
+    if( relic_data && relic_data->max_charges() > 0 && relic_data->charges_per_use() > 0 ) {
```
Comment (KorGgenT) [inline src/item.cpp:4679]:
```suggestion
    if( is_relic() && relic_data->max_charges() > 0 && relic_data->charges_per_use() > 0 ) {
```

**Comment by KorGgenT** [comment]
This is an excellent start! I want these things for the procgen in order to be as flexible as possible:
- A "range" of max charges (the artifact spawns with a max charges of between x and y)
- a range of time (the artifact can recharge every z time which is between x and y)
- a range of charges_per_use
- these ranges are ranges in the procgen_data only, the artifact should have a singular number
- All of these should have power level values
- give active_procgen_values a power level as well, per spell
- in relic::power_level() and relic_procgen_data::power_level() add all these up

I believe all of these features will make for very well-rounded procedural generation rules for artifacts.

File: src/relic.cpp [hunk 2]
```diff
@@ -160,6 +191,73 @@ void relic_procgen_data::deserialize( JsonIn &jsin )
...
+    jo.read( "recharge_type", type );
+    jo.read( "time", time );
+}
+
+relic_charge_info relic_charge_template::generate() const
+{
+    relic_charge_info ret;
+    ret.max_charges = rng( max_charges.first, max_charges.second );
+    ret.charges_per_use = rng( charges_per_use.first, charges_per_use.second );
+    ret.charges = rng( init_charges.first, init_charges.second );
```
Comment (KorGgenT) [inline src/relic.cpp:227]: two options
```
ret.charges = rng( init_charges.first, std::min( init_charges.second, ret.max_charges ) );
```
or
```
ret.charges = std::min( rng( init_charges.first, init_charges.second ), ret.max_charges );

**Comment by KorGgenT** [comment]
image
i don't know what's going on here.

**Comment by anothersimulacrum** [comment]
#42693

File: src/relic.cpp [hunk 3]
```diff
@@ -217,16 +317,81 @@ void relic::serialize( JsonOut &jsout ) const
...
 
+    jsout.member( "charge_info", charge );
+
     jsout.end_object();
 }
 
-int relic::activate( Creature &caster, const tripoint &target ) const
+int relic::activate( Creature &caster, const tripoint &target )
 {
+    if( charges() - charge.charges_per_use < 0 ) {
```
Comment (KorGgenT) [inline src/relic.cpp:327]:
```suggestion
    if( charge.charges_per_use != 0 && charges() - charge.charges_per_use < 0 ) {
```

**Pull request merged**